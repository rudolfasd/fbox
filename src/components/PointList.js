import React from 'react';
import { SortableContainer, SortableElement } from 'react-sortable-hoc';
import arrayMove from 'array-move';


const SortableItem = SortableElement(({id,num,text,delItemFunction}) => {
	return (
		<div className='item-wrap'>
			<div className='item'>
				<div className='item-num'>{num}</div>
				<div className='item-text'>
					{text}
				</div>
				<div className='item-remove'
					onClick={() => delItemFunction({id})}>
				</div>
			</div>
		</div>
	);
});

const SortableList = SortableContainer(({order,points,delItemFunction}) => {
	return(
		<div className='point-list-overflow'>
			<div className='point-list'>
				{order.map((id, index) => (
					<SortableItem
						key={`item-${id}`}
						index={index}
						id={id}
						num={index + 1}
						text={points[id]}
						delItemFunction={delItemFunction}
					/>
				))}
			</div>
		</div>
	);
});


class PointList extends React.Component {

	order = () => this.props.pointList.order;

	onSortEnd = ({oldIndex, newIndex}) => {
		const newOrderedList = arrayMove(this.order(), oldIndex, newIndex);
		this.props.sagaReorderList({order: newOrderedList});
	};
	shouldCancelStart = (e) => {
		let classList = e.target.classList;
		if (classList.contains('item-remove')) {
			return true;
		};
		return false;
	};
	render(){
		return (
				<SortableList
					lockAxis='y'
					lockToContainerEdges={true}
					helperClass='point-dragging'
					onSortEnd={this.onSortEnd}
					shouldCancelStart={this.shouldCancelStart}
					order={this.order()}
					points={this.props.pointList.points}
					delItemFunction={this.props.sagaDelPoint}
				/>
		);
	};
};

export { SortableItem, SortableList };
export default PointList;

