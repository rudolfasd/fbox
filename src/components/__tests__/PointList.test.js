import React from 'react';

import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import PointList from '../PointList';
import { SortableItem } from '../PointList';
import arrayMove from 'array-move';


jest.mock('array-move');


describe('REACT COMPONENT <PointList /> - Snapshots',() => {

    const mockSagaDelPoint = jest.fn();
    const mockSagaReorderList = jest.fn();
    const mockProps = { 
        pointList: {
            nextId: 9,
            order: [2,5,6,8],
            points: {
                2: 'qwe',
                5: 'asd',
                6: 'zxc',
                8: 'rty'
            }
        },
        sagaDelPoint: mockSagaDelPoint,
        sagaReorderList: mockSagaReorderList
    };
    let renderedValue, component;

    it('SortableItem snapshot', () => {
        renderedValue = renderer.create(
            <SortableItem 
                id={2} num={1} text={'Some text...'}
                delItemFunction={mockSagaDelPoint} 
            />).toJSON();
        expect(renderedValue).toMatchSnapshot();
    });
    it('SortableItem - onclick delete', () => {
        component = shallow(
            <SortableItem 
                id={2} num={1} text={'Some text...'}
                delItemFunction={mockSagaDelPoint} 
            />);
        component.find('.item-remove').first().simulate('click');
        expect(mockSagaDelPoint).toHaveBeenCalledTimes(1);
        expect(mockSagaDelPoint).toHaveBeenCalledWith({id: 2});
    });

    it('PointList snapshot', () => {
        renderedValue = renderer.create(
            <PointList {...mockProps}/>
        ).toJSON();
        expect(renderedValue).toMatchSnapshot();
    });
    it('PointList - SortableList onSortEnd', () => {
        arrayMove.mockImplementation(() => { return 'Reordered Array'});
        component = shallow(<PointList {...mockProps} />);
        component.instance().onSortEnd({oldIndex: 0, newIndex: 1});
        expect(arrayMove).toHaveBeenCalledTimes(1);
        expect(arrayMove).toHaveBeenCalledWith( [2,5,6,8],0,1 );
        expect(mockSagaReorderList).toHaveBeenCalledTimes(1);
        expect(mockSagaReorderList).toHaveBeenCalledWith({order: 'Reordered Array'});
    });
    it('PointList - SortableList shouldCancelStart', () => {
        const mockEventFromElementWithSomeClass = {
            target: {classList: {contains: (someClass) => true } }};
        const mockEventFromElementWithoutSomeClass = {
            target: {classList: {contains: (someClass) => false} }};
            component = shallow(<PointList {...mockProps} />);
        expect(component.instance().shouldCancelStart(mockEventFromElementWithSomeClass))
            .toEqual(true);
        expect(component.instance().shouldCancelStart(mockEventFromElementWithoutSomeClass))
            .toEqual(false); 
    });
});