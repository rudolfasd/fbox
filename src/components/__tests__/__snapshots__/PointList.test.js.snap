// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`REACT COMPONENT <PointList /> - Snapshots PointList snapshot 1`] = `
<div
  className="point-list-overflow"
>
  <div
    className="point-list"
  >
    <div
      className="item-wrap"
    >
      <div
        className="item"
      >
        <div
          className="item-num"
        >
          1
        </div>
        <div
          className="item-text"
        >
          qwe
        </div>
        <div
          className="item-remove"
          onClick={[Function]}
        />
      </div>
    </div>
    <div
      className="item-wrap"
    >
      <div
        className="item"
      >
        <div
          className="item-num"
        >
          2
        </div>
        <div
          className="item-text"
        >
          asd
        </div>
        <div
          className="item-remove"
          onClick={[Function]}
        />
      </div>
    </div>
    <div
      className="item-wrap"
    >
      <div
        className="item"
      >
        <div
          className="item-num"
        >
          3
        </div>
        <div
          className="item-text"
        >
          zxc
        </div>
        <div
          className="item-remove"
          onClick={[Function]}
        />
      </div>
    </div>
    <div
      className="item-wrap"
    >
      <div
        className="item"
      >
        <div
          className="item-num"
        >
          4
        </div>
        <div
          className="item-text"
        >
          rty
        </div>
        <div
          className="item-remove"
          onClick={[Function]}
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`REACT COMPONENT <PointList /> - Snapshots SortableItem snapshot 1`] = `
<div
  className="item-wrap"
>
  <div
    className="item"
  >
    <div
      className="item-num"
    >
      1
    </div>
    <div
      className="item-text"
    >
      Some text...
    </div>
    <div
      className="item-remove"
      onClick={[Function]}
    />
  </div>
</div>
`;
